File: fromEmidio/_mercuryapi-1.3.1.83/_mercuryapi-M6E_B0-1.3.1.83/cs/ThingMagic.Reader/Iso180006b.cs
/*
 * Copyright (c) 2009 ThingMagic, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
using System;
using System.Collections.Generic;

namespace ThingMagic
{
    /// <summary>
    /// Iso tag-specific constructs
    /// </summary>
    public static class Iso180006b
    {
        #region Nested Enums

	#region LinkFrequency

        /// <summary>
        /// enum to define link frequency options
        /// </summary>
	public enum LinkFrequency
        {
	  /// <summary>
	  /// 40kHz link frequency
	  /// </summary>
	  LINK40KHZ = 40,
	  /// <summary>
	  /// 160kHz link frequency
	  /// </summary>
	  LINK160KHZ = 160,
	}

        #endregion

        #region SelectOp

        /// <summary>
        /// enum to define select options
        /// </summary>
        public enum SelectOp
        {
            /// <summary>
            /// Select Tags that match the select criteria
            /// </summary>            
            EQUALS,
            /// <summary>
            /// Select Tags that do not match the select criteria
            /// </summary>
            NOTEQUALS,
            /// <summary>
            /// Select Tags that match select data that are less less than the select criteria
            /// </summary>
            LESSTHAN,
            /// <summary>
            /// Select Tags that match select data that are greated than the select criteria
            /// </summary>
            GREATERTHAN,
        }
 
        #endregion

        #endregion

        #region Nested Classes

        #region TagData

        /// <summary>
        /// ISO specific version of TagData.
        /// </summary>
        public class TagData : ThingMagic.TagData
        {
            #region Properties

            /// <summary>
            /// Tag's RFID protocol
            /// </summary>
            public override TagProtocol Protocol
            {
                get
                {
                    return TagProtocol.ISO180006B;
                }
            }

            #endregion

            #region Construction

            /// <summary>
            /// Create TagData with blank CRC
            /// </summary>
            /// <param name="epcBytes">EPC value</param>
            public TagData(ICollection<byte> epcBytes) : base(epcBytes) { }

            /// <summary>
            /// Create TagData
            /// </summary>
            /// <param name="epcBytes">EPC value</param>
            /// <param name="crcBytes">CRC value</param>
            public TagData(ICollection<byte> epcBytes, ICollection<byte> crcBytes) : base(epcBytes, crcBytes) { } 

            #endregion
        }

        #endregion

        #region LockAction
        /// <summary>
        /// Iso180006B lockAction
        /// </summary>
        public class LockAction : TagLockAction
        {
            #region Fields

            /// <summary>
            /// The address of the byte to lock.
            /// </summary>
            public byte Address;

            #endregion

            #region Construction

            /// <summary>
            /// Create a lock action for the given address
            /// </summary>
            /// <param name="address">Memory address to lock</param>
            public LockAction(byte address)
            {
                Address = address;
            }

	    #endregion

	}
        #endregion

        #region Select
        /// <summary>
        /// Representation of a ISO Select Operation.
        /// </summary>
        public class Select : TagFilter
        {
            #region Fields

            /// <summary>
            /// Whether to invert the selection (deselect tags that meet the comparison and vice versa).
            /// </summary>
            public bool Invert;

            /// <summary>
            /// The select option to provide
            /// </summary>
            public SelectOp Op;

            /// <summary>
            /// The select address
            /// </summary>
            public byte Address;

            /// <summary>
            /// A bitmask of which of the eight provided bytes to compare
            /// </summary>
            public byte Mask;

            /// <summary>
            /// The data to compare. Exactly eight bytes.
            /// </summary>
            public byte[] Data; 

            #endregion

            #region Construction

            /// <summary>
            /// Select constructor.
            /// </summary>
            /// <param name="invert">invert the selection</param>
            /// <param name="op">select options</param>
            /// <param name="address">starting mask address</param>
            /// <param name="mask">the select mask</param>
            /// <param name="data">the data for comparison</param>
            public Select(bool invert, SelectOp op, byte address, byte mask, ICollection<byte> data)
            {
                Invert = invert;
                Op = op;
                Address = address;
                Mask = mask;
                Data = CollUtil.ToArray(data);

                if (Data.Length != 8)
                    throw new ArgumentException("ISO180006B select data must be 8 bytes");
            }

            #endregion

            #region Matches

            /// <summary>
            /// Predicate for post-processing filters.
            /// </summary>
            /// <param name="t">tag data to screen</param>
            /// <returns>Return true to allow tag through the filter.
            /// Return false to reject tag.</returns>
            public bool Matches(ThingMagic.TagData t)
            {
                throw new NotSupportedException();
            }

            #endregion

            /// <summary>
            /// Returns a String that represents the current Object.
            /// </summary>
            /// <returns>A String that represents the current Object.</returns>
            public override string ToString()
            {
                return String.Format(
                    "Iso180006b.Select:[{0}{1},{2},0x{3:X2},{4}]",
                    (Invert ? "Invert," : ""),
                    Op, Address, Mask, ByteFormat.ToHex(Data));

            }
        }

        #endregion

        #region LockTag
        /// <summary>
        /// Perform a lock or unlock operation on a tag. The first tag seen
        /// is operated on - the singulation parameter may be used to control
        /// this. Note that a tag without an access password set may not
        /// accept a lock operation or remain locked.
        /// </summary>
        /// 
        public class LockTag : TagOp
        {
            #region Fields

            /// <summary>
            /// the locking action to take
            /// </summary>
            public TagLockAction Action;

            #endregion

            #region Construction
            /// <summary>
            /// Constructor to initialize the parameters of LockTag
            /// </summary>
            /// <param name="action">the locking action to take</param>
            /// 
            public LockTag(TagLockAction action)
            {
                this.Action = action;
            }

            #endregion

        }

        #endregion

        #region WriteTagMemWords
        /// <summary>
        /// Write data to the memory bank of a tag.
        /// </summary>
        /// 
        public class WriteTagMemWords : TagOp
        {
            #region Fields

            /// <summary>
            /// the byte address to start writing to
            /// </summary>
            public int Address;

            /// <summary>
            /// the bytes to write
            /// </summary>
            public ICollection<ushort> Data;

            #endregion

            #region Construction
            /// <summary>
            /// Constructor to initialize the parameters of WriteTagMemBytes
            /// </summary>
            /// <param name="address">the byte address to start writing to</param>
            /// <param name="data">the bytes to write</param>
            /// 
            public WriteTagMemWords(int address, ICollection<ushort> data)
            {
                this.Address = address;
                this.Data = data;
            }

            #endregion

        }

        #endregion

        #region ReadTagMemWords
        /// <summary>
        /// Read data from the memory bank of a tag.
        /// </summary>
        /// 
        public class ReadTagMemWords : TagOp
        {
            #region Fields

            /// <summary>
            /// the word address to start reading at
            /// </summary>
            public int WordAddress;

            /// <summary>
            /// the number of words to read
            /// </summary>
            public int WordCount;

            #endregion

            #region Construction
            /// <summary>
            /// Constructor to initialize the parameters of ReadTagMemBytes
            /// </summary>
            /// <param name="wordAddress">the word address to start reading at</param>
            /// <param name="wordCount">the number of words to read</param>
            /// 
            public ReadTagMemWords(int wordAddress, int wordCount)
            {
                this.WordAddress = wordAddress;
                this.WordCount = wordCount;
            }

            #endregion

        }

        #endregion

        

        #endregion

    }
}

File: fromEmidio/_mercuryapi-1.3.1.83/_mercuryapi-M6E_B0-1.3.1.83/java/mercuryapi_nb/src/com/thingmagic/Iso180006b.java
/*
 * Copyright (c) 2008 ThingMagic, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.thingmagic;

/**
 * This class is a namespace for ISO 18000-6B-specific subclasses of
 * generic Mercury API classes, data structures, constants, and
 * convenience functions.
 */
public class Iso180006b
{

  // non-public, this class is never instantiated
  Iso180006b() {}


  public enum LinkFrequency
  {
      LINK40KHZ (0),
      LINK160KHZ (1);

    int rep;
    LinkFrequency(int rep)
    {
      this.rep = rep;
    }
  }

/**
 * This class extends {@link TagData} to represent the details of an 
 * ISO 18000-6B RFID tag.
 */
  public static class TagData extends com.thingmagic.TagData
  {

    @Override
    boolean checkLen(int epcbytes)
    {
      if (epcbytes != 8)
      {
        return false;
      }
      return true;
    }

    public TagProtocol getProtocol()
    {
      return TagProtocol.ISO180006B;
    }

    /**
     * Construct an ISO 18000-6B tag data from a byte array.
     *
     * @param bEPC EPC bytes. Must be 8 bytes
     */
    public TagData(byte[] bEPC)
    {
      super(bEPC);
    }

    /**
     * Construct an ISO 18000-6B tag data from a byte array.
     *
     * @param bEPC EPC bytes. Must be 8 bytes
     * @param crc CRC bytes
     */
    public TagData(byte[] bEPC, byte[] crc)
    {
      super(bEPC, crc);
    }

    /**
     * Construct an ISO 18000-6B tag data from a hexadecimal string.
     *
     * @param sEPC Hex string. Must be 8 bytes (16 hex digits)
     */
    public TagData(String sEPC)
    {
      super(sEPC);
    }

    /**
     * Construct an ISO 18000-6B tag data from a hexadecimal string.
     *
     * @param sEPC Hex string. Must be 8 bytes (16 hex digits)
     * @param sCRC Hex string. Must be 2 bytes (4 hex digits)
     */
    public TagData(String sEPC, String sCRC)
    {
      super(sEPC, sCRC);
    }

  }


  /**
   * The arguments to a Reader.lockTag() method for ISO18000-6B tags. 
   * Represents the address of a byte of memory to lock
   */
  public static class LockAction
    extends TagLockAction
  {
    int address;

    /**
     * Construct a new LockAction from an address.
     *
     * @param address the address of the byte to lock. Must be in the range 0-255
     */
    public LockAction(int address)
    {
      if (address < 0 || address > 255)
      {
        throw new IllegalArgumentException("ISO18000 memory address out of range");
      }
      this.address = address;
    }
  }

  /**
   * Operations that can be performed as part of a Group Select operation
   */
  public enum SelectOp
  {
        /** Select if tag data matches op data */
      EQUALS      (0),
        /** Select if tag data does not match op data */
      NOTEQUALS   (1),
        /** Select if tag data is less than op data */
      LESSTHAN    (2),
        /** Select if tag data is greater than op data */
      GREATERTHAN (3);

    int rep; /* air protocol representation of the value */
    SelectOp(int rep)
    {
      this.rep = rep;
    }

  }
  public static class Select implements TagFilter
  {

    /**
     * Whether to invert the selection (deselect tags that meet the
     * comparison and vice versa).
     */
    public boolean invert;

    /**
     * The operation to compare the tag data to the provided data.
     */
    SelectOp op;

    /**
     * The address of the tag data to compare the the provided data.
     */ 
    int address;

    /**
     * A bitmask of which of the eight provided bytes to compare.
     */
    byte mask;
    
    /**
     * The data to compare. Exactly eight bytes.
     */
    byte[] data;

    public Select(boolean invert, SelectOp op,
                  int address, byte mask, byte[] data)
    {
      this.invert = invert;
      this.op = op;
      this.address = address;
      this.mask = mask;
      this.data = data.clone();
      if (this.data.length != 8)
      {
        throw new IllegalArgumentException("ISO180006B select data must be 8 bytes");
      }
    }

    public boolean matches(com.thingmagic.TagData t)
    {
      throw new UnsupportedOperationException();
    }

    @Override
    public String toString()
    {
      StringBuilder dataHex = new StringBuilder(data.length * 2);

      for (byte b : data)
      {
        dataHex.append(String.format("%02X", b));
      }
      
      return String.format("Iso180006b.Select:[%s%s,%d,%d,%s]",
                           invert ? "Invert," : "",
                           op, address, mask, 
                           dataHex.toString());
    }

  }
}
